include/basic_types.h

```diff
@@ -82,6 +82,7 @@ typedef uint8_t seL4_Uint8;
 typedef uint16_t seL4_Uint16;
 typedef uint32_t seL4_Uint32;
 typedef word_t seL4_Word;
+typedef uintcap_t seL4_Ptr;
```

Comment:
Absolutely not, it is not a pointer to a cap. Just use `seL4_Ptr` instead.

Reply:
Not sure I follow. `seL4_Ptr` doesn't mean it's a pointer to a cap. It's just a type that could hold a valid CHERI cap (when CHERI is enabled) or an integer. The convention and intention of each type I added is described in this [commit message,](https://github.com/seL4/seL4/pull/1322/commits/a467df787d0bff65ab7cc3a838946ac703ef6449) but I am happy to rename to something that makes more sense.

Reply:
We definitely can't use just `cap` in the CHERI sense in names unless it is absolutely clear that it is not an seL4 cap. We're mixing two capability systems and CHERI caps are only visible in some configs, so the word `cap` always should be referring to seL4 caps. Would it make sense to use `cheri_cap`? (I haven't checked where these are used, the PR doesn't seem finished enough yet). 

I'm not keen on having CHERI terminology in generic seL4 types, esp names such as `uintcap_t` which only make sense to people who know CHERI (and know it fairly well). I'm fine with more fine-grained types to make intentions clearer, but if they are used in generic code, they need to make sense to C programmes who have never seen CHERI.

Reply:
I am open to any naming suggestions. As the commit message suggests, this `uintcap_t` is going to work with or without CHERI and basically should be able to hold biggest register width (eg for resider types, IPC message registers, sys call args, etc). I thought of naming it `register_t` as well (but this is already taken by seL4 for register IDs). `cheri_cap` name wouldn't make sense when we're building for non-CHERI code.

Reply:
@lsf37 I tested building and testing the PR for both non-CHERI and CHERI (RISC-V and AArch64).

Reply:
> cheri_cap name wouldn't make sense when we're building for non-CHERI code.

Ok, that means any `*cap_t` doesn't make sense for non-CHERI code for this. Not sure. Some variation on `word_t` maybe that expresses a minimum width?

Reply:
> > cheri_cap name wouldn't make sense when we're building for non-CHERI code.
> 
> Ok, that means any `*cap_t` doesn't make sense for non-CHERI code for this. Not sure. Some variation on `word_t` maybe that expresses a minimum width?

`cap_t` is still seL4's capability. I haven't touched that. I am not sure I understand what you mean by `*cap_t` though.
`uintcap_t` (or whatever we're going to rename it to) is just a generic type that'll work for all seL4 variants, and will be able to hold a maximum register-width size. `word_t` is left to *only* represent integers, never pointers. In non-CHERI builds, `word_t` and `uintcap_t` will be the same (e.g., `unsigned long` for RV64)

Reply:
By `*cap_t` I just meant all new names that have `cap_t` in them (I've only seen `uintcap_t`, just wanted to indicate it's not specifically that one only). We really need to be careful with the word cap in this code base.

I would have expected `word_t` to be a generic type, possibly related to register words, and `uint_t` to be an integer type.

Reply:
I see, thanks for the clarification. It was a trade-off whether to change `word_t` to be of register size/format/type, or introduce a new type. When I did some prototyping with changing `word_t` to be of the register type, this was way more disruptive as seL4 used `word_t` for pointer, arithmetic, registers, integers, and pretty much everything. It was less disruptive and more intentional to retype the very core use cases and types that may contain CHERI capabilities, and leave `word_t` to be for integers. We can still rename all the remaining`word_t` to `uint_t` after this, now (part of this PR) we know for sure they're only used for integers, not pointers.  

I have now refactored naming a bit to avoid confusions and be more explicit in the following commit (I copy it here for convenience):

>  [cheri][shared][types] Decouple integer and pointer types
>     
>     In conventional C, integers and pointers have the same representation in
>     hardware and it is fine to use and cast both to each other interchangeably;
>     i.e., from (char *) to (unsigned long). CHERI C enforces memory safety
>     by decoupling integer from pointer types (which are hardware
>     capabilities with bounds and permissions). This change aims to 1)
>     comply with CHERI C, and 2) enhance intentionality and readability of
>     types in seL4, even without CHERI.
>     
>     regoff_t: This is a new type rename for the existing seL4 "register_t" type
>     used for holding register offsets, not actual register values. We renamed that
>     to regoff_t in this commit for more clarity and to avoid confusion,
>     and also to re-claim "register_t" name as below.
>     
>     register_t: newly added to *always* hold full harwdare-register-width variables.
>     In the current seL4 and CHERI support, this type will always have to be used
>     for holding:
>       1) GPR HW register context.
>       2) IPC message registers.
>       3) Generic system call arguments.
>     
>     Or any other variable that may, at any point, need to hold a full-width and
>     format HW register (eg, special HW capability registers).
>     
>      - In non-CHERI modes, this is just an integer and corresponds to
>     unsigned long or the legacy word_t type.
>      - In any CHERI mode, this is a capability-width (and format) type and
>     corresponds to CHERI's __uintcap_t. It could be viewed as a super type that is
>     able to hold normal integers, CHERI pointers/capabilities (valid or invalid).
>     
>     seL4_Register is the user-level/libsel4 version of the kernel's register_t.
>     
>     uintptr_t: newly added and is used here for any variable that may hold a
>     capability or just an integer or some arithmetic is done on it. Used
>     generically if the variable could hold pointers in purecap mode and
>     it's not known what it could hold from the other types below.
>      - In non-CHERI modes, this is just an integer and corresponds to
>     unsigned long.
>      - In hybrid CHERI mode, this is also an integer and corresponds to
>     unsigned long.
>      - In purecap CHERI mode, this is a capability-width type and
>     corresponds to CHERI's __uintcap_t.
>     
>     ptraddr_t: newly added and is defined to represent the address
>     portion of the pointer. This is basically an integer.
>     
>     size_t: Defined to represent unsigned sizes, also an integer.
>     
>     word_t: is used by seL4 as some type that can hold anything (eg
>     unsigned long, and is the most widely used for mixed use cases such as
>     pointers, integers, registers, etc. It's left it as it is to only
>     represent "integers".
>     
>     pptr_t: conventionally only holds kernel addresses. Same as uintptr_t
>     vptr_t: conventionally only holds user addresses. Same as uintptr_t

Reply:
> My advise is to minimise changes as much as reasonably possible.
> 
Agreed. I view adding CHERI support (especially purecap in the kernel) as not only to enforce extra security, but also to have more robust coding and typing conventions, intentionality, and readability for non-CHERI code as well. So it's a good opportunity to do things right with such a big feature, if it makes it to seL4. 

> > I see, thanks for the clarification. It was a trade-off whether to change `word_t` to be of register size/format/type, or introduce a new type. When I did some prototyping with changing `word_t` to be of the register type, this was way more disruptive as seL4 used `word_t` for pointer, arithmetic, registers, integers, and pretty much everything. It was less disruptive and more intentional to retype the very core use cases and types that may contain CHERI capabilities, and leave `word_t` to be for integers.
> 
> Agreed.
> 
> > We can still rename all the remaining`word_t` to `uint_t` after this, now (part of this PR) we know for sure they're only used for integers, not pointers.
> 
> I wouldn't bother, as that is so widespread, including in libsel4, that it's a lost cause by now.
> 
> > > regoff_t: This is a new type rename for the existing seL4 "register_t" type
> > > used for holding register offsets, not actual register values. We renamed that
> > > to regoff_t in this commit for more clarity and to avoid confusion,
> > > and also to re-claim "register_t" name as below.
> 
> Is this necessary? Can't we use an existing type for this, like word_t or sword_t?
> 
> > > register_t: newly added to _always_ hold full harwdare-register-width variables.
> 
> Agreed.
> 
> > > In the current seL4 and CHERI support, this type will always have to be used
> > > for holding:
> > > 
> > > 1. GPR HW register context.
> 
> Yes.
> 
> > > 2. IPC message registers.
> > > 3. Generic system call arguments.
> 
> Disagreed with those two, as should be clear by my other posts.
> 
I'd have to stick with my design/implementation decision for the reasons I mentioned in this [comment](https://github.com/seL4/seL4/pull/1322#issuecomment-2371349677).

Just to be extra clear, I completely agree with you that no valid pointers/capabilities should be passed between different VSpaces (address spaces). Enforcing that is another issue. But the IPC message registers aren't only used for inter-address-space IPC, and this change doesn't really have to do with enforcing/prohibiting transferring "valid" CHERI capabilities between address spaces. 

> > > seL4_Register is the user-level/libsel4 version of the kernel's register_t.
> 
> Fine.
> 
> > > uintptr_t: newly added and is used here for any variable that may hold a
> > > capability or just an integer or some arithmetic is done on it. Used
> > > generically if the variable could hold pointers in purecap mode and
> > > it's not known what it could hold from the other types below.
> 
> Not sure whether this one is needed: seL4 shouldn't be stuffing pointers in integers for its own use. Just use register_t for the few cases you need this.
> 
For Morello/RV64, `register_t` will be 128-bit when we build the kernel in hybrid mode and userspace in purecap mode. `uintptr_t` and friends will need to be 64-bit regardless, in that same setup, in the kernel. 

> > > ptraddr_t: newly added and is defined to represent the address
> > > portion of the pointer. This is basically an integer.
> 
> We already have paddr_t.
> 
`paddr_t` is for physical integers. `ptraddr_t` is for when we're dealing with the "address part" of the CHERI capability, whether that's for user, kernel, etc. This is a newly added type in CHERI projects for intentionality and clarity, and other CHERI-based systems use it that way.

> > > size_t: Defined to represent unsigned sizes, also an integer.
> 
> Why add it?
Not really biased, I could remove, but I think it'd be useful for good practices, readability, and intentionality (similar to most other C projects) in seL4 to use to hold "sizes" instead of just using `word_t` which could be confusing what it's defined to. Also the `%z` printf specifier would work with this size more probably than `word_t`

It's similar to having pptr_t, vptr_t, word_t, etc all `unsigned long`, but for better coding conventions we have them and should really be using them in the right places all over the code, instead of mixing using them with `word_t`.

Reply:
> `paddr_t` is for physical integer addresses. `ptraddr_t` is for when we're dealing with the "address part" of the CHERI capability, whether that's for user, kernel, etc. This is a newly added type in CHERI projects for intentionality and clarity, and other CHERI-based systems use it that way.

`ptraddr_t` is too similar to `paddr_t` in name. If you're dealing with a virtual address, then use `vptr_t` instead.

Reply:
`vptr_t` is conventionally just used for user "pointers". So it can't be used here. `ptraddr_t` is not a pointer and can't be, it's the "integer address" part of the "CHERI pointer" which is basically just the same as `word_t`